apps/src/music/MusicLabView.jsx

```diff
@@ -0,0 +1,257 @@
+import React, {useEffect, useState, useRef, useCallback} from 'react';
+import PropTypes from 'prop-types';
+import {useFetch} from '../util/useFetch';
+import AnalyticsReporter from './analytics/AnalyticsReporter';
+import MusicBlocklyWorkspace from './blockly/MusicBlocklyWorkspace';
+import {AnalyticsContext, BaseUrlContext, PlayerUtilsContext} from './context';
+import MusicPlayer from './player/MusicPlayer';
+import StatelessMusicView from './StatelessMusicView';
+import {getStore} from '@cdo/apps/redux';
+import {connect, Provider} from 'react-redux';
+import queryString from 'query-string';
+import {SignInState} from '@cdo/apps/templates/currentUserRedux';
+import Globals from './globals';
+import {setCurrentAudioElapsedTime, setSoundEvents} from './musiclabRedux';
+
+const DEFAULT_GROUP_NAME = 'all';
+const baseUrl = 'https://curriculum.code.org/media/musiclab/';
+
+const UnconnectedMusicLabView = ({
+  userId,
+  userType,
+  signInState,
+  setCurrentAudioElapsedTime,
+  setSoundEvents
+}) => {
+  // Refs
+  const playerRef = useRef(new MusicPlayer());
+  const reporterRef = useRef(new AnalyticsReporter());
+  const workspaceRef = useRef(new MusicBlocklyWorkspace());
+  const timerIdRef = useRef(null);
+
+  // State
+  const [isPlaying, setIsPlaying] = useState(false);
+  const [allSounds, setAllSounds] = useState([]);
+
+  let parameters = queryString.parse(location.search);
+  const libraryFilename = parameters['library']
+    ? `music-library-${parameters['library']}.json`
+    : 'music-library.json';
+
+  const {data} = useFetch(baseUrl + libraryFilename, null, [parameters]);
+  const library = data;
+
+  // const startTimer = () => {
+  //   stopTimer();
+  //   timerIdRef.current = setInterval(
+  //     () =>
+  //       setCurrentAudioElapsedTime(
+  //         playerRef.current.getCurrentAudioElapsedTime()
+  //       ),
+  //     1000 / 30
+  //   );
+  // };
+
+  // const stopTimer = () => {
+  //   if (timerIdRef.current) {
+  //     clearInterval(timerIdRef.current);
+  //     timerIdRef.current = null;
+  //   }
+  // };
+
+  const clearCode = () => {
+    workspaceRef.current.resetCode();
+
+    stopSong();
+
+    playerRef.current.clearAllSoundEvents();
+  };
+
+  const executeSong = () => {
+    workspaceRef.current.executeSong({
+      MusicPlayer: playerRef.current
+    });
+    setSoundEvents(playerRef.current.getSoundEvents());
+  };
+
+  const playSong = () => {
+    playerRef.current.stopSong();
+
+    // Clear the events list of when_run sounds, because it will be
+    // populated next.
+    playerRef.current.clearWhenRunEvents();
+    executeSong();
+    playerRef.current.playSong();
+    setIsPlaying(true);
+
+    console.log('playSong', Blockly.getWorkspaceCode());
```

Comment:
It feels like we now have a lot of business logic in the view that could instead be moved into the reducer, here and in other actions in this view.